### PR TITLE
Run rustdoc-gui tests in parallel

### PR DESCRIPTION
**Pull request by dns2utf8**
I hid the passing tests and only show the failed ones in alphabetical order:
image

Also this PR cuts down the execution time from ~40 to ~9 seconds

### CODE REVIEW

**Comment by dns2utf8** [comment]
I extracted this from #86594 with @GuillaumeGomez

File: src/tools/rustdoc-gui/tester.js [hunk 1]
```diff
@@ -70,6 +70,14 @@ function parseOptions(args) {
     return null;
 }
 
+function print_test_successful() {
+    process.stdout.write(".");
+}
+function print_test_erroneous() {
```
Comment (GuillaumeGomez) [inline src/tools/rustdoc-gui/tester.js:76]: NIt: please add an empty line before this function.

File: src/tools/rustdoc-gui/tester.js [hunk 2]
```diff
@@ -70,6 +70,14 @@ function parseOptions(args) {
...
+    // Bold Red "F" Reset
+    process.stdout.write("\033[1m\x1b[31mF\x1b[0m");
```
Comment (GuillaumeGomez) [inline src/tools/rustdoc-gui/tester.js:78]: That might work well for linux, but not on windows at all. ;)

File: src/tools/rustdoc-gui/tester.js [hunk 3]
```diff
@@ -100,27 +108,48 @@ async function main(argv) {
...
-        await runTest(testPath, options).then(out => {
+        tests.push(runTest(testPath, options));
+    }
+
+    let error_outputs = "";
+    let failed_outputs = "";
+    for (var i = 0; i < tests.length; ++i) {
+        await tests[i].then(out => {
             const [output, nb_failures] = out;
-            console.log(output);
```
Comment (GuillaumeGomez) [inline src/tools/rustdoc-gui/tester.js:114]: Instead of removing this, why simply storing all output into an array and then sort it once done before displaying it? Like that it'll remain sorted and everything.

File: src/tools/rustdoc-gui/tester.js (hunk 2 above)
Comment (dns2utf8) [reply src/tools/rustdoc-gui/tester.js:78]: I wished that JavsScript had `#[cfg(unix)]` ...

File: src/tools/rustdoc-gui/tester.js [hunk 4]
```diff
@@ -100,27 +108,48 @@ async function main(argv) {
...
-            console.error(err);
+            error_outputs += err + "\n";
+            print_test_erroneous();
             failed = true;
         });
     }
+    console.log("")
+
     if (failed) {
+        console.log(failed_outputs);
```
Comment (GuillaumeGomez) [inline src/tools/rustdoc-gui/tester.js:151]: In case you enabled debug and your test is passing, you still want this output.

File: src/tools/rustdoc-gui/tester.js [hunk 5]
```diff
@@ -100,27 +108,48 @@ async function main(argv) {
...
             if (nb_failures > 0) {
+                failed_outputs += output + "\n";
+                print_test_erroneous()
                 failed = true;
+            } else {
+                print_test_successful()
             }
         }).catch(err => {
-            console.error(err);
+            error_outputs += err + "\n";
```
Comment (GuillaumeGomez) [inline src/tools/rustdoc-gui/tester.js:143]: Instead of putting everything into one string, it would maybe be better to keep an array of the tests and have some info for each, like output, errors, etc...

File: src/tools/rustdoc-gui/tester.js [hunk 6]
```diff
@@ -100,27 +108,48 @@ async function main(argv) {
...
+    let tests = [];
     if (opts["files"].length === 0) {
         files = fs.readdirSync(opts["tests_folder"]).filter(file => path.extname(file) == ".goml");
     } else {
         files = opts["files"].filter(file => path.extname(file) == ".goml");
     }
-
+    if (files.length === 0) {
+        console.log("rustdoc-gui: No test selected");
+        process.exit(2);
```
Comment (GuillaumeGomez) [inline src/tools/rustdoc-gui/tester.js:119]: Nice add!

**Comment by GuillaumeGomez** [comment]
I also think we should disable running checks in parallel when `--no-headless` is used (take a look at `src/test/rustdoc-gui/README.md` to see how to use it.

File: src/tools/rustdoc-gui/tester.js (hunk 2 above)
Comment (GuillaumeGomez) [reply src/tools/rustdoc-gui/tester.js:78]: If that's the only thing you wished js had, I guess it's fine. 😉 

For now, just keep it simple and simply print text.

File: src/tools/rustdoc-gui/tester.js (hunk 4 above)
Comment (dns2utf8) [reply src/tools/rustdoc-gui/tester.js]: That is a good argument

File: src/tools/rustdoc-gui/tester.js (hunk 2 above)
Comment (dns2utf8) [reply src/tools/rustdoc-gui/tester.js:79]: I found an easy way to detect linux at runtime and enable the colour, what do you think?

File: src/tools/rustdoc-gui/tester.js (hunk 6 above)
Comment (dns2utf8) [reply src/tools/rustdoc-gui/tester.js:124]: Thanks :)

**Comment by dns2utf8** [comment]
I implemented sequential execution if the `--no-headless` flag is present.

However, I was unable to make it run with `./x.py`:
```bash
./x.py test --stage 1 --jobs 8 --no-headless --debug rustdoc-gui

Updating only changed submodules
Submodules updated in 0.03 seconds
    Finished dev [unoptimized + debuginfo] target(s) in 0.15s

Unrecognized option: 'no-headless'

Usage: x.py <subcommand> [options] [<paths>...]
...
```

File: src/tools/rustdoc-gui/tester.js (hunk 3 above)
Comment (dns2utf8) [reply src/tools/rustdoc-gui/tester.js:114]: see main comment

File: src/tools/rustdoc-gui/tester.js (hunk 4 above)
Comment (dns2utf8) [reply src/tools/rustdoc-gui/tester.js]: Done

**Comment by dns2utf8** [comment]
Regarding ordering the logs, I perallocate an Array now that the async tests fill with a result object. This can happen in order with the `--no-headless` flag or out of order without it.

The output then remains sorted with the exception of execution errors, which are at the end because I wanted people working with them to see them at the end of the output where they should be simpler to spot.
That just an idea I had and we can change it to just sorted output

File: src/tools/rustdoc-gui/tester.js (hunk 2 above)
Comment (GuillaumeGomez) [reply src/tools/rustdoc-gui/tester.js:79]: I
 think it's completely beyond scope. I'd rather keep it as simple as possible. Please simply print the text (on stderr if it's an error).

File: src/tools/rustdoc-gui/tester.js [hunk 7]
```diff
@@ -101,25 +118,91 @@ async function main(argv) {
...
+    results.forEach(r => {
+        switch (r.status) {
+            case RUN_SUCCESS:
+            case RUN_FAILED:
+                // skip
+                break;
+            case RUN_ERRORED:
+                console.error(r.output);
+                break;
+            default:
```
Comment (GuillaumeGomez) [inline src/tools/rustdoc-gui/tester.js:200]: This default case has already been checked above, so it's unnecessary here.

File: src/tools/rustdoc-gui/tester.js [hunk 8]
```diff
@@ -101,25 +118,91 @@ async function main(argv) {
...
+        }
     }
+    await Promise.all(tests);
+    // final \n after the tests
+    console.log("\n");
+
+    results.forEach(r => {
+        switch (r.status) {
+            case RUN_SUCCESS:
+                if (debug === false) {
```
Comment (GuillaumeGomez) [inline src/tools/rustdoc-gui/tester.js:176]: You're supposed to display that the test was run.

File: src/tools/rustdoc-gui/tester.js [hunk 9]
```diff
@@ -101,25 +118,91 @@ async function main(argv) {
...
-            const [output, nb_failures] = out;
-            console.log(output);
-            if (nb_failures > 0) {
+        tests.push(
+            runTest(testPath, options)
+            .then(out => {
+                //console.log(i);
+                const [output, nb_failures] = out;
+                results[i] = {
+                    status: nb_failures === 0 ? RUN_SUCCESS : RUN_FAILED,
```
Comment (GuillaumeGomez) [inline src/tools/rustdoc-gui/tester.js:146]: Instead of this, why not have three arrays, one for success, one for failures and one for errors? It even makes it simpler to handle it afterwards.

File: src/tools/rustdoc-gui/tester.js (hunk 8 above)
Comment (GuillaumeGomez) [reply src/tools/rustdoc-gui/tester.js:176]: Ah nevermind, I see what you did! You used the "." display like it's used for other tests. That's a nice idea!

**Comment by GuillaumeGomez** [comment]
> I implemented sequential execution if the `--no-headless` flag is present.
> 
> However, I was unable to make it run with `./x.py`:
> 
> ```shell
> ./x.py test --stage 1 --jobs 8 --no-headless --debug rustdoc-gui
> 
> Updating only changed submodules
> Submodules updated in 0.03 seconds
>     Finished dev [unoptimized + debuginfo] target(s) in 0.15s
> 
> Unrecognized option: 'no-headless'
> 
> Usage: x.py <subcommand> [options] [<paths>...]
> ...
> ```

Look [here](https://github.com/rust-lang/rust/blob/master/src/test/rustdoc-gui/README.md) how to pass arguments to the tester through `x.py`.

File: src/tools/rustdoc-gui/tester.js (hunk 7 above)
Comment (dns2utf8) [reply src/tools/rustdoc-gui/tester.js:200]: True

File: src/tools/rustdoc-gui/tester.js (hunk 9 above)
Comment (dns2utf8) [reply src/tools/rustdoc-gui/tester.js:143]: Hm, I guess my mind was very rusty and I emulated an enum :upside_down_face:

Comment (GuillaumeGomez) [reply src/tools/rustdoc-gui/tester.js:143]: ;)

File: src/bootstrap/test.rs [hunk 10]
```diff
@@ -834,22 +834,68 @@ impl Step for RustdocJSNotStd {
     }
 }
 
-fn check_if_browser_ui_test_is_installed_global(npm: &Path, global: bool) -> bool {
+fn get_browser_ui_test_version_from_string(input: &String) -> Option<String> {
```
Comment (GuillaumeGomez) [inline src/bootstrap/test.rs:837]: Same, this is out of scope for this PR. Please open another one for checking the package version.

Comment (dns2utf8) [reply src/bootstrap/test.rs:837]: Sure, will do :+1:

File: src/tools/rustdoc-gui/tester.js (hunk 9 above)
Comment (dns2utf8) [reply src/tools/rustdoc-gui/tester.js:143]: Should we keep it as is? With the separate Arrays I would still have to pre-allocate and then filter afterwards.

File: src/tools/rustdoc-gui/tester.js [hunk 11]
```diff
@@ -70,12 +70,24 @@ function parseOptions(args) {
     return null;
 }
 
+function print_test_successful() {
+    process.stdout.write(".");
+}
+
+function print_test_erroneous() {
+    process.stdout.write("F");
```
Comment (GuillaumeGomez) [inline src/tools/rustdoc-gui/tester.js:78]: `console.log` wouldn't do it here?

File: src/tools/rustdoc-gui/tester.js [hunk 12]
```diff
@@ -70,12 +70,24 @@ function parseOptions(args) {
     return null;
 }
 
+function print_test_successful() {
+    process.stdout.write(".");
```
Comment (GuillaumeGomez) [inline src/tools/rustdoc-gui/tester.js:74]: `console.log` wouldn't do it here?

File: src/tools/rustdoc-gui/tester.js [hunk 13]
```diff
@@ -101,25 +115,88 @@ async function main(argv) {
...
     if (opts["files"].length === 0) {
-        files = fs.readdirSync(opts["tests_folder"]).filter(file => path.extname(file) == ".goml");
+        files = fs.readdirSync(opts["tests_folder"]);
     } else {
-        files = opts["files"].filter(file => path.extname(file) == ".goml");
+        files = opts["files"];
+    }
+    files = files.filter(file => path.extname(file) == ".goml");
+    if (files.length === 0) {
+        console.log("rustdoc-gui: No test selected");
```
Comment (GuillaumeGomez) [inline src/tools/rustdoc-gui/tester.js:124]:
```suggestion
        console.error("rustdoc-gui: No test selected");
```

File: src/tools/rustdoc-gui/tester.js [hunk 14]
```diff
@@ -101,25 +115,88 @@ async function main(argv) {
...
+        process.exit(2);
     }
-
     files.sort();
-    for (var i = 0; i < files.length; ++i) {
+
+    console.log(`running ${files.length} rustdoc-gui tests`);
```
Comment (GuillaumeGomez) [inline src/tools/rustdoc-gui/tester.js:129]:
```suggestion
    console.log(`Running ${files.length} rustdoc-gui tests...`);
```

File: src/tools/rustdoc-gui/tester.js [hunk 15]
```diff
@@ -101,25 +115,88 @@ async function main(argv) {
...
                 failed = true;
-            }
-        }).catch(err => {
-            console.error(err);
-            failed = true;
-        });
+            })
+        );
+        if (no_headless) {
+            await tests[i];
```
Comment (GuillaumeGomez) [inline src/tools/rustdoc-gui/tester.js:163]: Nice! :+1:

Comment (GuillaumeGomez) [reply src/tools/rustdoc-gui/tester.js:163]: My only concern here is that now we don't know which test is running anymore though...

File: src/tools/rustdoc-gui/tester.js [hunk 16]
```diff
@@ -101,25 +115,88 @@ async function main(argv) {
...
+    results.forEach(r => {
+        switch (r.status) {
+            case RUN_SUCCESS:
+            case RUN_FAILED:
+                // skip
+                break;
+            case RUN_ERRORED:
+                console.error(r.output);
+                break;
+        }
```
Comment (GuillaumeGomez) [inline src/tools/rustdoc-gui/tester.js:197]: I think you can simply rewrite it as follows:

```suggestion
        if (r.status === RUN_ERRORED) {
            console.error(r.output);
        }
```

File: src/tools/rustdoc-gui/tester.js (hunk 11 above)
Comment (dns2utf8) [reply src/tools/rustdoc-gui/tester.js:78]: No, because console.log adds a '\n' to each output

File: src/tools/rustdoc-gui/tester.js (hunk 12 above)
Comment (dns2utf8) [reply src/tools/rustdoc-gui/tester.js:74]: No, because console.log adds a '\n' to each output

Comment (GuillaumeGomez) [reply src/tools/rustdoc-gui/tester.js:74]: Oh right! Good point.

File: src/tools/rustdoc-gui/tester.js [hunk 17]
```diff
@@ -101,25 +115,88 @@ async function main(argv) {
...
+    for (let i = 0; i < files.length; ++i) {
         const testPath = path.join(opts["tests_folder"], files[i]);
-        await runTest(testPath, options).then(out => {
-            const [output, nb_failures] = out;
-            console.log(output);
-            if (nb_failures > 0) {
+        tests.push(
+            runTest(testPath, options)
+            .then(out => {
+                //console.log(i);
```
Comment (GuillaumeGomez) [inline src/tools/rustdoc-gui/tester.js:140]: Artifact? :)

File: src/tools/rustdoc-gui/tester.js (hunk 9 above)
Comment (GuillaumeGomez) [reply src/tools/rustdoc-gui/tester.js:143]: I'm not sure there is much point to pre-allocate anything. Maybe simply keep the promise around and push it to the "correct" array once it's done?

File: src/tools/rustdoc-gui/tester.js (hunk 15 above)
Comment (dns2utf8) [reply src/tools/rustdoc-gui/tester.js:163]: I just ran the test with this command:
```bash
./x.py test src/test/rustdoc-gui --stage 1 --jobs 8 --test-args --no-headless
```
For me it is very hard to tell what is going on because the windows pop up, do something for a second or two and then flash away.
Should the dev confirm each run?

Comment (dns2utf8) [reply src/tools/rustdoc-gui/tester.js:163]: Sorry, each test

File: src/tools/rustdoc-gui/tester.js (hunk 16 above)
Comment (dns2utf8) [reply src/tools/rustdoc-gui/tester.js:197]: Oh, yes that is better :+1:

File: src/tools/rustdoc-gui/tester.js (hunk 17 above)
Comment (dns2utf8) [reply src/tools/rustdoc-gui/tester.js]: Ah, yes :)

File: src/tools/rustdoc-gui/tester.js (hunk 9 above)
Comment (dns2utf8) [reply src/tools/rustdoc-gui/tester.js:142]: Then to keep the order we would have to keep the initial index and sort all the result arrays

File: src/tools/rustdoc-gui/tester.js (hunk 2 above)
Comment (dns2utf8) [reply src/tools/rustdoc-gui/tester.js]: Done :+1:

File: src/tools/rustdoc-gui/tester.js (hunk 9 above)
Comment (GuillaumeGomez) [reply src/tools/rustdoc-gui/tester.js:142]: That sounds fine to me. :p

File: src/tools/rustdoc-gui/tester.js (hunk 15 above)
Comment (GuillaumeGomez) [reply src/tools/rustdoc-gui/tester.js:162]: Actually it doesn't matter. I guess we're only interesting in the one failing.

File: src/tools/rustdoc-gui/tester.js (hunk 9 above)
Comment (GuillaumeGomez) [reply src/tools/rustdoc-gui/tester.js:142]: Or you can store the test name in the array too. Like: `{name: test_name, output: output}` and then you sort on the `name` field.

**Comment by JohnCSimon** [comment]
Ping from triage:
@dns2utf8 Can you please post your status on this PR? Is it ready to be reviewed? If so set `S-waiting-on-reviewer`

**Comment by dns2utf8** [comment]
I implemented the result capturing with three arrays and the output now looks a lot like the other testers
```
Running 25 rustdoc-gui tests...
.......... (10/25)
.......... (20/25)
.....
```

[asciicast](https://asciinema.org/a/JGdSqlW4eyAU8lOLH4fsexbq4)

File: src/tools/rustdoc-gui/tester.js [hunk 18]
```diff
@@ -101,25 +136,77 @@ async function main(argv) {
...
-        }).catch(err => {
-            console.error(err);
-            failed = true;
+            })
+        );
+        if (no_headless) {
+            await tests[i];
+        }
+    }
+    await Promise.all(tests);
```
Comment (GuillaumeGomez) [inline src/tools/rustdoc-gui/tester.js:190]: Shouldn't it be:

```suggestion
if (!no_headless) {
       await Promise.all(tests);
}
```

?

File: src/tools/rustdoc-gui/tester.js [hunk 19]
```diff
@@ -101,25 +136,77 @@ async function main(argv) {
...
+    // final \n after the tests
+    console.log("\n");
```
Comment (GuillaumeGomez) [inline src/tools/rustdoc-gui/tester.js:192]: Shouldn't it be:

```suggestion
    console.log("");
```

? Or do you specifically want an empty line?

Comment (GuillaumeGomez) [reply src/tools/rustdoc-gui/tester.js:192]: If so, maybe better to use two `console.log("")` instead I think. No strong opinion though.

**Review by GuillaumeGomez — COMMENTED** [review]
A few questions to be answered but otherwise it looks really promising!

File: src/tools/rustdoc-gui/tester.js (hunk 18 above)
Comment (dns2utf8) [reply src/tools/rustdoc-gui/tester.js:190]: Logic wise kind of yes. But you can await a future multiple times, the work will get done only once

File: src/tools/rustdoc-gui/tester.js (hunk 19 above)
Comment (dns2utf8) [reply src/tools/rustdoc-gui/tester.js:192]: I wanted a bit more separation between the sections for readability so the progress bar is visually spaced from the outputs.
The output should be the same on all platforms because we don't write anything on the empty line but fully correct it should be two `console.log("")`, will change that

Comment (dns2utf8) [reply src/tools/rustdoc-gui/tester.js:192]: Replaced it with a better finish method that prints the final stats

**Comment by dns2utf8** [comment]
I am using `os.EOL` now for the line endings. Also the status_bar now has a finish function that aligns the last progress bar. 

Here is an example run with two failed tests and `--debug`:
image

And here is one without:
image

One small question for me is how many characters do we want per line? Currently I picked 10, but this is arbitrary and maybe a bit low?

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
> One small question for me is how many characters do we want per line? Currently I picked 10, but this is arbitrary and maybe a bit low?

Other test-suites use 100 but I don't care. We can always update it later. Looks all good to me now, thanks a lot!

@bors: r+

**Comment by Mark-Simulacrum** [comment]
Is this respecting -j passed to x.py somehow? We should avoid running multiple processes if -j1 is passed, for example.

**Comment by GuillaumeGomez** [comment]
@Mark-Simulacrum Good point. I'll open an issue for it.
